File: backend/routes/insights.py
"""
Splito Smart Insights Engine — pure sqlite3
GET /api/insights              — personal insight cards
GET /api/insights/group/<id>   — group-level insights
"""

from flask import Blueprint, jsonify
from database import get_conn
from utils.auth import require_auth
from collections import Counter, defaultdict
from datetime import datetime, timezone, timedelta

insights_bp = Blueprint('insights', __name__)


# ── Helpers ────────────────────────────────────────────────────────────────

def _now():
    return datetime.now(timezone.utc)


def _my_group_ids(user_id):
    with get_conn() as conn:
        rows = conn.execute(
            "SELECT group_id FROM group_members WHERE user_id=?", (user_id,)
        ).fetchall()
    return [r['group_id'] for r in rows]


def _group_name(group_id):
    with get_conn() as conn:
        r = conn.execute("SELECT name FROM groups_ WHERE id=?", (group_id,)).fetchone()
    return r['name'] if r else group_id


def _group_expenses(group_id, days=30):
    since = (_now() - timedelta(days=days)).isoformat()
    with get_conn() as conn:
        return conn.execute("""
            SELECT e.*, u.username as paid_by_name
            FROM expenses e JOIN users u ON u.id = e.paid_by_id
            WHERE e.group_id=? AND e.created_at >= ?
        """, (group_id, since)).fetchall()


# ── Individual insight generators ─────────────────────────────────────────

def _insight_repeated_payer(group_id, group_name, user_id):
    with get_conn() as conn:
        exps = conn.execute("""
            SELECT paid_by_id FROM expenses
            WHERE group_id=? ORDER BY created_at DESC LIMIT 10
        """, (group_id,)).fetchall()
    if len(exps) < 3:
        return None

    counts = Counter(r['paid_by_id'] for r in exps)
    top_id, top_cnt = counts.most_common(1)[0]
    ratio = top_cnt / len(exps)
    if ratio < 0.6:
        return None

    with get_conn() as conn:
        u = conn.execute("SELECT username FROM users WHERE id=?", (top_id,)).fetchone()
    payer_name = u['username'] if u else 'Someone'

    if top_id == user_id:
        msg    = f"You've paid for {top_cnt} of the last {len(exps)} expenses in {group_name}. It's time for others to chip in! 💸"
        action = "Send a reminder"
    else:
        msg    = f"{payer_name} has covered {top_cnt} of the last {len(exps)} expenses in {group_name}. It might be your turn next!"
        action = "Offer to pay"

    return {
        'type': 'repeated_payer', 'priority': 'high', 'icon': '🔄',
        'title': 'Repeated Payer Detected', 'message': msg, 'action': action,
        'group_id': group_id, 'group_name': group_name,
        'data': {'payer_id': top_id, 'payer_name': payer_name, 'ratio': round(ratio * 100)},
    }


def _insight_overspending(group_id, group_name):
    exps = _group_expenses(group_id, days=30)
    if not exps:
        return None

    total = sum(e['amount'] for e in exps)
    cats  = defaultdict(float)
    for e in exps:
        cats[e['category']] += e['amount']

    for cat, amt in cats.items():
        if total > 0 and amt / total >= 0.40 and amt > 500:
            pct = round(amt / total * 100)
            return {
                'type': 'overspending', 'priority': 'medium', 'icon': '⚠️',
                'title': f'High {cat} Spend',
                'message': f'{pct}% of {group_name}\'s budget went to {cat} this month (₹{amt:,.0f}). Consider setting a limit.',
                'action': 'View breakdown',
                'group_id': group_id, 'group_name': group_name,
                'data': {'category': cat, 'amount': amt, 'percentage': pct},
            }
    return None


def _insight_weekly_summary(group_id, group_name):
    exps = _group_expenses(group_id, days=7)
    if not exps:
        return None
    total   = sum(e['amount'] for e in exps)
    top_cat = Counter(e['category'] for e in exps).most_common(1)[0][0]
    return {
        'type': 'weekly_summary', 'priority': 'low', 'icon': '📊',
        'title': 'Weekly Snapshot',
        'message': f'{group_name} spent ₹{total:,.0f} this week across {len(exps)} expense(s). Top category: {top_cat}.',
        'action': 'See analytics',
        'group_id': group_id, 'group_name': group_name,
        'data': {'total': total, 'count': len(exps), 'top_category': top_cat},
    }


def _insight_balance_warning(group_id, group_name, user_id):
    from routes.groups import compute_balances
    bal    = compute_balances(group_id)
    my_bal = bal.get(user_id, 0)
    if my_bal >= -100:
        return None
    return {
        'type': 'balance_warning', 'priority': 'high', 'icon': '💰',
        'title': 'Unsettled Balance',
        'message': f'You owe ₹{abs(my_bal):,.0f} in {group_name}. Settling up keeps friendships intact! 😊',
        'action': 'Settle now',
        'group_id': group_id, 'group_name': group_name,
        'data': {'balance': my_bal},
    }


def _insight_big_spender(group_id, group_name):
    exps = _group_expenses(group_id, days=7)
    if not exps:
        return None
    biggest = max(exps, key=lambda e: e['amount'])
    if biggest['amount'] < 500:
        return None
    return {
        'type': 'big_spender', 'priority': 'low', 'icon': '🏆',
        'title': 'Biggest Expense This Week',
        'message': f'"{biggest["title"]}" was {group_name}\'s biggest purchase at ₹{biggest["amount"]:,.0f}.',
        'action': 'View expense',
        'group_id': group_id, 'group_name': group_name,
        'data': {'expense_id': biggest['id'], 'amount': biggest['amount']},
    }


def _insight_savings_streak(group_id, group_name):
    """Celebrate if this week's spend is lower than last week's."""
    this_week = sum(e['amount'] for e in _group_expenses(group_id, days=7))
    last_week_exps = []
    with get_conn() as conn:
        since2 = (_now() - timedelta(days=14)).isoformat()
        till   = (_now() - timedelta(days=7)).isoformat()
        last_week_exps = conn.execute("""
            SELECT amount FROM expenses
            WHERE group_id=? AND created_at >= ? AND created_at < ?
        """, (group_id, since2, till)).fetchall()
    last_week = sum(e['amount'] for e in last_week_exps)
    if last_week > 0 and this_week < last_week * 0.8:
        saved = round(last_week - this_week, 2)
        return {
            'type': 'savings_streak', 'priority': 'low', 'icon': '🌱',
            'title': 'Spending Down This Week!',
            'message': f'{group_name} spent ₹{saved:,.0f} less than last week. Great discipline!',
            'action': 'Keep it up',
            'group_id': group_id, 'group_name': group_name,
            'data': {'saved': saved},
        }
    return None


# ── Personal stats ─────────────────────────────────────────────────────────

def _personal_stats(user_id, group_ids):
    since = (_now() - timedelta(days=30)).isoformat()
    with get_conn() as conn:
        paid_total = conn.execute("""
            SELECT COALESCE(SUM(amount),0) as total
            FROM expenses WHERE paid_by_id=? AND created_at >= ?
        """, (user_id, since)).fetchone()['total']

        fav_rows = conn.execute("""
            SELECT category, COUNT(*) as cnt
            FROM expenses WHERE paid_by_id=? AND created_at >= ?
            GROUP BY category ORDER BY cnt DESC LIMIT 1
        """, (user_id, since)).fetchone()

    fav_cat = fav_rows['category'] if fav_rows else 'N/A'

    from routes.groups import compute_balances
    owed_to_me = 0.0
    i_owe      = 0.0
    for gid in group_ids:
        bal = compute_balances(gid)
        b   = bal.get(user_id, 0)
        if b > 0:  owed_to_me += b
        else:      i_owe      += abs(b)

    return {
        'total_paid_30d':   round(paid_total, 2),
        'total_owed_to_me': round(owed_to_me, 2),
        'total_i_owe':      round(i_owe, 2),
        'net_balance':      round(owed_to_me - i_owe, 2),
        'favourite_category': fav_cat,
        'active_groups':    len(group_ids),
    }


# ── Routes ─────────────────────────────────────────────────────────────────

@insights_bp.route('', methods=['GET'])
@require_auth
def get_insights(current_user):
    uid       = current_user['id']
    group_ids = _my_group_ids(uid)
    all_ins   = []

    for gid in group_ids:
        gname = _group_name(gid)
        for fn in (_insight_repeated_payer, _insight_balance_warning):
            try:
                ins = fn(gid, gname, uid)
                if ins: all_ins.append(ins)
            except Exception: pass

        for fn in (_insight_overspending, _insight_weekly_summary,
                   _insight_big_spender, _insight_savings_streak):
            try:
                ins = fn(gid, gname)
                if ins: all_ins.append(ins)
            except Exception: pass

    priority_order = {'high': 0, 'medium': 1, 'low': 2}
    all_ins.sort(key=lambda x: priority_order.get(x['priority'], 3))

    return jsonify({
        'insights':       all_ins[:8],
        'personal_stats': _personal_stats(uid, group_ids),
    })


@insights_bp.route('/group/<group_id>', methods=['GET'])
@require_auth
def group_insights(current_user, group_id):
    with get_conn() as conn:
        if not conn.execute(
            "SELECT 1 FROM group_members WHERE group_id=? AND user_id=?",
            (group_id, current_user['id'])
        ).fetchone():
            return jsonify({'error': 'Not a member'}), 403

    gname  = _group_name(group_id)
    uid    = current_user['id']
    result = []
    for fn in (_insight_overspending, _insight_weekly_summary,
               _insight_big_spender, _insight_savings_streak):
        try:
            ins = fn(group_id, gname)
            if ins: result.append(ins)
        except Exception: pass
    for fn in (_insight_repeated_payer, _insight_balance_warning):
        try:
            ins = fn(group_id, gname, uid)
            if ins: result.append(ins)
        except Exception: pass

    return jsonify({'insights': result})

File: backend/routes/insights.js
const express = require('express');
const router  = express.Router();
const db   = require('../database');
const auth = require('../middleware/auth');

// GET /api/insights/summary — overall balance totals
router.get('/summary', auth, (req, res) => {
  // We need to compute total precise balances globally using the same mechanism as group balances.
  // 1. Get all expenses
  const splits = db.prepare(`
    SELECT es.amount, e.paid_by, es.user_id, e.group_id
    FROM expense_splits es
    JOIN expenses e ON e.id = es.expense_id
    JOIN group_members gm ON gm.group_id = e.group_id AND gm.user_id = ?
  `).all(req.userId);

  // 2. Get all settlements
  const settlements = db.prepare(`
    SELECT s.amount, s.from_user, s.to_user, s.group_id
    FROM settlements s
    JOIN group_members gm ON gm.group_id = s.group_id AND gm.user_id = ?
    WHERE s.is_paid = 1
  `).all(req.userId);

  // Calculate pairwise balances: balances[groupId][userId] = net amount the other user owes 'me' inside that group
  // Actually, we can just compute pairwise balances globally.
  // pair_balances[other_user_id] = net balance
  const pairBalances = {};

  splits.forEach(s => {
    if (s.paid_by === req.userId && s.user_id !== req.userId) {
      pairBalances[s.user_id] = (pairBalances[s.user_id] || 0) + s.amount;
    }
    if (s.user_id === req.userId && s.paid_by !== req.userId) {
      pairBalances[s.paid_by] = (pairBalances[s.paid_by] || 0) - s.amount;
    }
  });

  settlements.forEach(s => {
    if (s.to_user === req.userId && s.from_user !== req.userId) {
      pairBalances[s.from_user] = (pairBalances[s.from_user] || 0) - s.amount;
    }
    if (s.from_user === req.userId && s.to_user !== req.userId) {
      pairBalances[s.to_user] = (pairBalances[s.to_user] || 0) + s.amount;
    }
  });

  let owedToMe = 0;
  let iOwe = 0;

  for (const amount of Object.values(pairBalances)) {
    if (amount > 0.01) owedToMe += amount;
    if (amount < -0.01) iOwe += Math.abs(amount);
  }

  // Total spent this month
  const now = new Date();
  const monthStart = `${now.getFullYear()}-${String(now.getMonth()+1).padStart(2,'0')}-01`;
  const monthlySpend = db.prepare(`
    SELECT COALESCE(SUM(e.amount), 0) AS total
    FROM expenses e
    JOIN group_members gm ON gm.group_id = e.group_id AND gm.user_id = ?
    WHERE e.paid_by = ? AND e.created_at >= ?
  `).get(req.userId, req.userId, monthStart).total;

  // Active groups
  const groupCount = db.prepare('SELECT COUNT(*) AS c FROM group_members WHERE user_id=?').get(req.userId).c;

  // Pending count
  const pendingCount = db.prepare(`
    SELECT COUNT(*) AS c FROM settlements WHERE from_user=? AND is_paid=0
  `).get(req.userId).c;

  res.json({
    owed_to_me: Math.round(owedToMe * 100) / 100,
    i_owe:      Math.round(iOwe * 100) / 100,
    net:        Math.round((owedToMe - iOwe) * 100) / 100,
    monthly_spend: Math.round(monthlySpend * 100) / 100,
    group_count:   groupCount,
    pending_count: pendingCount,
  });
});

// GET /api/insights/categories — spending breakdown by category
router.get('/categories', auth, (req, res) => {
  const { period = 'month' } = req.query;
  const since = getSinceDate(period);

  const rows = db.prepare(`
    SELECT e.category, COALESCE(SUM(es.amount), 0) AS total, COUNT(*) AS count
    FROM expense_splits es
    JOIN expenses e ON e.id = es.expense_id
    JOIN group_members gm ON gm.group_id = e.group_id AND gm.user_id = ?
    WHERE es.user_id = ? AND e.created_at >= ?
    GROUP BY e.category
    ORDER BY total DESC
  `).all(req.userId, req.userId, since);

  const grand = rows.reduce((s, r) => s + r.total, 0);
  res.json(rows.map(r => ({
    category: r.category,
    total:    Math.round(r.total * 100) / 100,
    count:    r.count,
    pct:      grand > 0 ? Math.round((r.total / grand) * 1000) / 10 : 0,
  })));
});

// GET /api/insights/trends — daily spending for graph
router.get('/trends', auth, (req, res) => {
  const { period = 'week' } = req.query;
  const since = getSinceDate(period);
  const fmt   = period === 'year' ? '%Y-%m' : '%Y-%m-%d';

  const rows = db.prepare(`
    SELECT strftime('${fmt}', e.created_at) AS date,
           COALESCE(SUM(es.amount), 0) AS total
    FROM expense_splits es
    JOIN expenses e ON e.id = es.expense_id
    JOIN group_members gm ON gm.group_id = e.group_id AND gm.user_id = ?
    WHERE es.user_id = ? AND e.created_at >= ?
    GROUP BY date
    ORDER BY date ASC
  `).all(req.userId, req.userId, since);

  res.json(rows.map(r => ({
    date:   r.date,
    amount: Math.round(r.total * 100) / 100,
    label:  fmtLabel(r.date, period),
  })));
});

// GET /api/insights/breakdown — COMBINED categories + trend for the Insights page
router.get('/breakdown', auth, (req, res) => {
  const { period = 'week' } = req.query;
  const since = getSinceDate(period);
  const fmt   = period === 'year' ? '%Y-%m' : '%Y-%m-%d';

  // Categories
  const catRows = db.prepare(`
    SELECT e.category, COALESCE(SUM(es.amount), 0) AS total
    FROM expense_splits es
    JOIN expenses e ON e.id = es.expense_id
    JOIN group_members gm ON gm.group_id = e.group_id AND gm.user_id = ?
    WHERE es.user_id = ? AND e.created_at >= ?
    GROUP BY e.category ORDER BY total DESC
  `).all(req.userId, req.userId, since);

  const grand = catRows.reduce((s, r) => s + r.total, 0);

  // Trend
  const trendRows = db.prepare(`
    SELECT strftime('${fmt}', e.created_at) AS date,
           COALESCE(SUM(es.amount), 0) AS total
    FROM expense_splits es
    JOIN expenses e ON e.id = es.expense_id
    JOIN group_members gm ON gm.group_id = e.group_id AND gm.user_id = ?
    WHERE es.user_id = ? AND e.created_at >= ?
    GROUP BY date ORDER BY date ASC
  `).all(req.userId, req.userId, since);

  // Prev period for comparison
  const prevSince = getSinceDate(period, 2);
  const prevTotal = db.prepare(`
    SELECT COALESCE(SUM(es.amount), 0) AS t
    FROM expense_splits es JOIN expenses e ON e.id=es.expense_id
    JOIN group_members gm ON gm.group_id=e.group_id AND gm.user_id=?
    WHERE es.user_id=? AND e.created_at>=? AND e.created_at<?
  `).get(req.userId, req.userId, prevSince, since).t;

  let comparison = null;
  if (prevTotal > 0) {
    const delta = Math.round(((grand - prevTotal) / prevTotal) * 100);
    comparison = delta < 0
      ? `You spent ${Math.abs(delta)}% less than last ${period}. 🎉 Great job saving!`
      : `You spent ${delta}% more than last ${period}. Consider reviewing ${catRows[0]?.category || 'your top'} expenses.`;
  }

  res.json({
    categories: catRows.map(r => ({
      category: r.category,
      amount:   Math.round(r.total * 100) / 100,
      pct:      grand > 0 ? Math.round((r.total / grand) * 1000) / 10 : 0,
    })),
    trend: trendRows.map(r => ({
      date:   r.date,
      amount: Math.round(r.total * 100) / 100,
      label:  fmtLabel(r.date, period),
    })),
    total: Math.round(grand * 100) / 100,
    comparison,
  });
});

// GET /api/insights/group-comparison — your spend vs group avg by category
router.get('/group-comparison', auth, (req, res) => {
  const { group_id, period = 'month' } = req.query;
  const since = getSinceDate(period);

  const query = (userId) => db.prepare(`
    SELECT e.category, COALESCE(SUM(es.amount), 0) AS total
    FROM expense_splits es
    JOIN expenses e ON e.id = es.expense_id
    ${group_id ? 'WHERE e.group_id=? AND' : 'JOIN group_members gm ON gm.group_id=e.group_id AND gm.user_id=? WHERE'} es.user_id=? AND e.created_at>=?
    GROUP BY e.category
  `).all(...(group_id ? [group_id, userId, since] : [userId, userId, since]));

  const mySpend    = Object.fromEntries(query(req.userId).map(r => [r.category, r.total]));
  const categories = [...new Set(Object.keys(mySpend))];

  res.json(categories.map(cat => ({
    category: cat,
    you:      Math.round((mySpend[cat] || 0) * 100) / 100,
  })));
});

function getSinceDate(period, multiplier = 1) {
  const d = new Date();
  if (period === 'week')  d.setDate(d.getDate() - 7 * multiplier);
  if (period === 'month') d.setMonth(d.getMonth() - 1 * multiplier);
  if (period === 'year')  d.setFullYear(d.getFullYear() - 1 * multiplier);
  return d.toISOString().slice(0, 10);
}

function fmtLabel(dateStr, period) {
  const d = new Date(dateStr + 'T00:00:00');
  if (period === 'year') return d.toLocaleString('en-IN', { month: 'short' });
  if (period === 'month') return d.toLocaleString('en-IN', { day: 'numeric', month: 'short' });
  return d.toLocaleString('en-IN', { weekday: 'short' }).slice(0,2);
}

module.exports = router;
